chain/network/src/network_protocol/proto_conv.rs

```diff
@@ -0,0 +1,528 @@
+use crate::network_protocol::proto;
+use crate::network_protocol::proto::peer_message::MessageType as ProtoMT;
+use crate::network_protocol::{
+    Handshake, HandshakeFailureReason, PeerMessage, RoutingSyncV2, RoutingTableUpdate,
+};
+use anyhow::{bail, Context};
+/// Contains protobuf <-> network_protocol conversions.
+use borsh::{BorshDeserialize as _, BorshSerialize as _};
+use near_network_primitives::types::{
+    Edge, PartialEdgeInfo, PeerChainInfoV2, PeerInfo, RoutedMessage,
+};
+use near_primitives::block::{Block, BlockHeader, GenesisId};
+use near_primitives::challenge::Challenge;
+use near_primitives::hash::CryptoHash;
+use near_primitives::network::{AnnounceAccount, PeerId};
+use near_primitives::syncing::{EpochSyncFinalizationResponse, EpochSyncResponse};
+use near_primitives::transaction::SignedTransaction;
+use near_primitives::types::EpochId;
+
+impl TryFrom<&proto::CryptoHash> for CryptoHash {
+    type Error = anyhow::Error;
+    fn try_from(p: &proto::CryptoHash) -> anyhow::Result<Self> {
+        CryptoHash::try_from(&p.hash[..]).map_err(|err| anyhow::Error::msg(err.to_string()))
+    }
+}
+
+impl TryFrom<&CryptoHash> for proto::CryptoHash {
+    type Error = anyhow::Error;
+    fn try_from(x: &CryptoHash) -> anyhow::Result<Self> {
+        Ok(Self { hash: x.0.into() })
+    }
+}
+
+//////////////////////////////////////////
+
+impl TryFrom<&proto::GenesisId> for GenesisId {
+    type Error = anyhow::Error;
+    fn try_from(p: &proto::GenesisId) -> anyhow::Result<Self> {
+        Ok(Self {
+            chain_id: p.chain_id.clone(),
+            hash: (|| CryptoHash::try_from(p.hash.as_ref().context("missing")?))()
+                .context("hash")?,
```

Comment:
In the context of how this context could be presented to the user, I think these messages should be expanded to be more descriptive than a single word. Perhaps something like

* the `hash` field in the `GenesisId` value is missing
* could not convert `CryptoHash` / could not parse `CryptoHash` [^1]

in this specific case. More descriptive messages would also obviate the need for the `(||)()` pattern.

[^1]: note that the error context should be describing the operation that occurs in the fallible expression, not the surrounding code. That is, in `hash: foo.context(...)?` the context should be describing what kind of operation `foo` is, and not `hash: foo.context(...)?` or even a broader context. In practice not maintaining a specific polarity in adding context can lead to situations where you end up with both missing context and duplicated context. While not a problem here, contextualizing strictly the failing expression also helps to avoid guessing the callers' intent.